src/lib.rs

```diff
@@ -41,1346 +53,988 @@ use std::{
-    /// state of the Fiat-Shamir rng.
-    /// TODO: rename this function
-    fn check_combinations_individual_opening_challenges<'a>(
-        vk: &Self::VerifierKey,
-        linear_combinations: impl IntoIterator<Item = &'a LinearCombination<F>>,
-        commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
-        eqn_query_set: &QuerySet<F>,
-        eqn_evaluations: &Evaluations<F>,
-        proof: &BatchLCProof<F, Self>,
+    /// Build the multi-poly single-point (random) linear combination.
+    // TODO: let us declare this function as internal.
+    fn single_point_multi_poly_open_combine<'a>(
+        labeled_polynomials: impl IntoIterator<Item = &'a LabeledPolynomial<G::ScalarField>>,
+        labeled_commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
         fs_rng: &mut Self::RandomOracle,
-    ) -> Result<bool, Self::Error>
-    where
-        Self::Commitment: 'a,
-    {
-        let BatchLCProof { proof, evals } = proof;
+        labeled_randomnesses: impl IntoIterator<Item = &'a LabeledRandomness<Self::Randomness>>,
+    ) -> Result<
+        (
+            Polynomial<G::ScalarField>, // the random linear combination of the p_i(X)
+            Self::Randomness,           // their combined commitment randomnesses.
+            bool, // `is_hiding` for the batch of polynomial. Is `true` if one of the p_i(X) needs hiding.
+        ),
+        Self::Error,
+    > {
+        let combine_time = start_timer!(|| "Combining polynomials, randomness, and commitments.");
+
+        let mut combined_polynomial = Polynomial::zero();
+        let mut combined_randomness = Self::Randomness::zero();
+
+        let mut has_hiding = false;
+
+        let polys_iter = labeled_polynomials.into_iter();
+        let rands_iter = labeled_randomnesses.into_iter();
+        let comms_iter = labeled_commitments.into_iter();
+
+        // as the statement/assertion of the opening proof is already bound to the interal state
+        // of the fr_rng, we simply squeeze the challenge scalar for the random linear combination
+        let lambda: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
+        let mut cur_challenge = G::ScalarField::one();
+
+        // compute the random linear combination using the powers of lambda
+        for (labeled_polynomial, (labeled_commitment, labeled_randomness)) in
+            polys_iter.zip(comms_iter.zip(rands_iter))
+        {
+            if labeled_polynomial.label() != labeled_commitment.label() {
+                end_timer!(combine_time);
+                Err(Error::Other(
+                    format!(
+                        "Labels are not equal: poly '{}'. commitment '{}'",
+                        labeled_polynomial.label(),
+                        labeled_commitment.label()
+                    )
+                    .to_owned(),
+                ))?
+            }
```

Comment:
It seems to me a bit strange: we pass the commitments only for checking them of the same order as "their" polynomials. Why don't we simply demand this per default?